src/eigen/lobpcg_hyper_impl.jl

```diff
@@ -43,17 +43,83 @@
 vprintln(args...) = nothing
 
 using LinearAlgebra
-using BlockArrays # used for the `mortar` command which makes block matrices
+import Base: *
+include("../workarounds/gpu_arrays.jl")
 
-# when X or Y are BlockArrays, this makes the return value be a proper array (not a BlockArray)
-function array_mul(X::AbstractArray{T}, Y) where T
-    Z = Array{T}(undef, size(X, 1), size(Y, 2))
-    mul!(Z, X, Y)
+# For now, BlockMatrix can store arrays of different types (for example, an element 
+# of type views and one of type Matrix). Maybe for performance issues it should only
+# store arrays of the same type?
+
+struct BlockMatrix
+    blocks::Tuple
```

Comment:
This is not type stable, which matters here because of performance. You have to make `Tuple` a concrete type.

Reply:
Yes, I thought this would also be an issue. The thing is, we sometimes build a BlockMatrix with arrays which are views of other arrays. And taking the view of an array builds a `SubArray` which derives from AbstractArray.
So mixed in BlockMatrix we can have either arrays or subarrays which do not have a concrete type in common: the only thing they have in common is that they derive from AbstractArray. I may be wrong, but I can't think of an easy way to make this field concrete by keeping the code the way it is: maybe we shouldn't give a BlockMatrix views?

Reply:
Like this?
```julia
struct BlockMatrix{T <: Tuple}
    blocks::T
end
```

Reply:
btw. Not sure we should have the `size` as a member. It's easily computed on the fly from the constituents of `blocks`.

Reply:
You're right: I was trying to describe more precisely what was hidden behind the Tuple, but doing `D <: Tuple` will work.
As for the size, I created this member so we don't have to recompute every time the size of the BlockMatrix. As it turns out, we are not using it much and it can be computed very easily, so we can drop it.